wait_all.go

```diff
@@ -0,0 +1,86 @@
+package asynctask
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+// WaitAllOptions defines options for WaitAll function
+type WaitAllOptions struct {
+	// FailFast set to true will indicate WaitAll to return on first error it sees.
+	FailFast bool
+}
+
+// WaitAll block current thread til all task finished.
+// first error from any tasks passed in will be returned.
+func WaitAll(ctx context.Context, options *WaitAllOptions, tasks ...*TaskStatus) error {
+	tasksCount := len(tasks)
+	errorCh := make(chan error, tasksCount)
+	errorChClosed := false
+	defer close(errorCh)
+	mutex := sync.Mutex{}
+
+	for _, tsk := range tasks {
+		go func(tsk *TaskStatus) {
```

Comment:
**suggestion**: extract the goroutine into a separate func for readability. maybe func waitOne(tsk *TaskStatus)